tidy: only run eslint on JS files if specified with "--extra-checks"

yotamofek opened the pull request:
`npx` already obviates the need to have the package installed locally, we can just ask `npx` to run the pre-configured version of `eslint`

This should reduce some friction for `rustdoc` contributors who are using `tidy`.
The one downside I can think of is that this will make *skipping* the `eslint` check more difficult to do, but not sure there's a good use-case for that.

r? @GuillaumeGomez since you added the `eslint` check to `tidy` :)

yotamofek committed:
tidy: only run `eslint` on JS files if specified with "--extra-checks"

GuillaumeGomez commented:
Seems good to me, thanks! r=me once CI pass.

@bors delegate+

bors commented:
:v: @yotamofek, you can now approve this pull request!

If @GuillaumeGomez told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@GuillaumeGomez`
<!-- homu: {"type":"Delegated","delegator":"GuillaumeGomez","delegate":"yotamofek"} -->

bors commented:
:pushpin: Commit 986b27087630e796180581c7de79398348c82a68 has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 986b27087630e796180581c7de79398348c82a68 -->
<!-- homu: {"type":"Approved","sha":"986b27087630e796180581c7de79398348c82a68","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Mark-Simulacrum commented:
@bors r-

How much extra time does this add for the tidy execution for folks who do have npx? We've had numerous attempts to optimize this flow, and I think the majority of contributors aren't editing JavaScript so this is just wasted time for them.

Maybe we can make this an opt-in in bootstrap.toml or perhaps detect changed files or similar?

yotamofek commented:
> > The one downside I can think of is that this will make skipping the eslint check more difficult to do, but not sure there's a good use-case for that.
> 
> ...I am not modifying the JS files in this repository.

Yeah, sorry, didn't phrase that correctly at all. I meant that not having `eslint` globally (or locally) installed isn't really a proper way to opt out of this check, since it's a kinda arbitrary thing that might be dependent on whether you by chance have some other JS project you're working on. NVM.

Your modified title isn't quite right, this PR doesn't change the way `eslint` is invoked (which was already via `npx`), just the check that is performed before.

yotamofek commented:
> @bors r-
> 
> How much extra time does this add for the tidy execution for folks who do have npx? We've had numerous attempts to optimize this flow, and I think the majority of contributors aren't editing JavaScript so this is just wasted time for them.
> 
> Maybe we can make this an opt-in in bootstrap.toml or perhaps detect changed files or similar?

This PR actually seems to make `tidy` run faster on my machine, I think asking `npm` to generate the list of installed packages was quite expensive.

But yeah, making the `eslint` opt-in (or only run when actually relevant) is probably a good idea.

workingjubilee commented:
> Your modified title isn't quite right, this PR doesn't change the way `eslint` is invoked (which was already via `npx`), just the check that is performed before.

Oh, well, yes, but the step that is `npm install` is replaced by simply directly running `npx eslint`, it just is replacing it like...

`(if (npm install package) then (npx package)) => (npx package)`

The original title just seemed slightly vague. Feel free to modify the PR title further or even revert it to your original if you disagree, I won't touch it again.

jieyouxu commented:
Doing this as part of `--extra-checks` also has an additional benefit -- I would expect `./x test tidy --extra-checks=js:lint` to hard-error if it cannot find a suitable `npx` or `eslint` locally, because that is now an opt-in flow that then contributor explicitly indicated their intent for.

jieyouxu commented:
Also hold on, will this try to use *any* eslint version, and not a pinned version?

workingjubilee commented on src/tools/tidy/src/rustdoc_js.rs:
```diff
@@ -9,9 +9,9 @@ use ignore::DirEntry;
+fn run_eslint(eslint_version: &str, args: &[PathBuf], config_folder: PathBuf, bad: &mut bool) {
     let mut child = match Command::new("npx")
-        .arg("eslint")
+        .arg(format!("eslint@{eslint_version}"))
```
@jieyouxu It should always try to use this version but I am not sure what the dependency resolution process for npx is from here, e.g. if it will use a compatible version that isn't the specified one.

yotamofek commented:
![image](https://github.com/user-attachments/assets/3a54fa02-2d47-4200-9c87-0d35ceb3560e)
Haha. Weird bug. I love it when GH contradicts itself 😁

yotamofek commented:
@jieyouxu mind taking a look please?
Moved the invocation of `eslint` to be gated by "--extra-checks"

yotamofek committed:
Add "js" to tidy extra checks for "mingw-check-tidy" CI job

Also, stop pre-installing `eslint` - the docker image is not cached in CI, so there's no point in explicitly `npm install`ing it instead of just letting `npx` download it

GuillaumeGomez commented:
Please don't. I don't want to gate eslint checks behind an option, too easy to miss.

jieyouxu commented:
I don't feel strongly enough to block running this in the default flow (I guess I could just uninstall `node` from my env too), but I would prefer if we gate eslint

- Behind `--extra-checks=js` (like is currently done), or
- Behind a bootstrap option, or
- Behind modified js files somehow.

Otherwise, you get penalized for having node installed even if you don't work on the js files...

yotamofek commented:
> Then let's keep it as it is currently? If `eslint` is not installed at the right version, then nothing happens. It's longer and a bit subpar but it prevents all listed issues from here.

Sounds good.
I'm contemplating between simply closing this PR and adding some comment in the code about why we're requiring `eslint` to be installed (e.g. "it's a poor man's opting-out mechanism") to prevent the next person here from getting nerd sniped like me.

GuillaumeGomez commented:
Adding comments and even a link to this PR would be a very good idea!

Kobzol commented:
I agree with @jieyouxu, detecting the version is not a good approach. It is not intuitive, to disable the check, you have to uninstall eslint, which you might want to have for other reasons, and if you have a different version, it fails tidy. The Python linting in tidy instead manages a virtualenv for you and installs the right version of packages, which is what we should do. Also, if you get a linting error once on CI, and then you install the correct eslint version locally, you'll then get worse tidy performance going forward.

We already have strong precedent for gating Python/C++ checks behind an optional flag. I don't see why we should do an exception for JS, it's in a similar category. For people who work on JS often, you can just invoke tidy with the extra check or write a bash script that will do it for you. With the opt-in, tidy can actually error out when eslint is missing on CI, which it currently doesn't; if eslint disappears from the CI image, we will silently stop running the linter now, which is not great. We will also stop checking the installed version of eslint, which is a hack anyway, and it makes running tidy slower.

GuillaumeGomez commented:
Well, kinda. We already use this mechanism in stuff like https://github.com/rust-lang/rust/pull/142677

Kobzol commented:
> No idea about the further improvements though, does origin/master get automatically updated by default in git pull somehow now?

We actually completely removed the remote handling, because it was unreliable and requires CI hacks. Now we just find the most recent commit authored by bors, that seems to be much more robust.

lolbinarycat commented:
> We actually completely removed the remote handling, because it was unreliable and requires CI hacks. Now we just find the most recent commit authored by bors, that seems to be much more robust.

just got `fmt: formatted 6117 modified files`, though I am unsure if this is because I modified tidy or something.

lolbinarycat commented:
Here's my attempt: https://github.com/rust-lang/rust/pull/142924
